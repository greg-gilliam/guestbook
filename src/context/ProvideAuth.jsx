import React, { createContext, useState } from 'react';

export const authContext = createContext();

export function ProvideAuth({ children }) {
  const [user, setUser] = useState(null);

  const login = (username, password) => {
    const loginSuccess =
      username === process.env.AUTH_USERNAME &&
      password === process.env.AUTH_PASSWORD;

    if (loginSuccess) setUser({ username });
    return loginSuccess;
  };

  const logout = (callback) => {
    setUser(null);
    callback();
  };

  return (
    <authContext.Provider value={{ user, login, logout }}>
      {children}
    </authContext.Provider>
  );
}
