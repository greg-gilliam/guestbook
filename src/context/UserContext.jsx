import { createContext, useContext, useState } from 'react';

const UserContext = createContext();

const UserProvider = ({ children }) => {
  const [user, setUser] = useState('');

  return (
    <UserContext.Provider value={{ user, setUser }}>
      {children}
    </UserContext.Provider>
  );
};

const useUser = () => {
  const context = useContext(UserContext);
  console.log('$$$', context);
  if (context === undefined) {
    throw new Error(
      'useUser hook must be called within a UserContext Provider'
    );
  }
  return context;
};

export { UserProvider, useUser };
