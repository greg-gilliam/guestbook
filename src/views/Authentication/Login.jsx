import React, { useState } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { useAuth } from '../../hooks/useAuth';
import { useUser } from '../../context/UserContext';
import { useForm } from '../../hooks/useForm';

export default function Login() {
  const history = useHistory();
  const location = useLocation();
  const auth = useAuth();
  const { setUser } = useUser();
  const { formState, handleFormChange } = useForm({
    username: '',
    password: '',
  });
  const [error, setError] = useState(null);

  const { from } = location.state || { from: { pathname: '/' } };

  const handleLogin = (event) => {
    event.preventDefault();
    const loginSuccess = auth.login(formState.username, formState.password);

    if (loginSuccess) {
      setUser(formState.username);
      history.replace(from.pathname);
    } else {
      setError('Invalid username and/or password');
    }
  };

  return (
    <fieldset>
      <legend>Sign in</legend>
      <form onSubmit={handleLogin}>
        <label htmlFor="username">Username</label>
        <input
          value={formState.username}
          onChange={handleFormChange}
          id="username"
          type="text"
          name="username"
        />
        <label htmlFor="password">Password</label>
        <input
          value={formState.password}
          onChange={handleFormChange}
          id="password"
          type="password"
          name="password"
        />
        <button type="submit" aria-label="sign in">
          Sign in!
        </button>
      </form>
      {error && <h3>{error}</h3>}
    </fieldset>
  );
}
