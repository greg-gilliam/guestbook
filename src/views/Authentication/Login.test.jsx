import { render } from '@testing-library/react';
import Login from './Login';
import { MemoryRouter, Switch, Route } from 'react-router-dom';
import { UserProvider } from '../../context/UserContext';
import { ProvideAuth } from '../../context/ProvideAuth';
import React from 'react';

it('should render the login page', () => {
  const { container } = render(
    <MemoryRouter>
      <UserProvider>
        <ProvideAuth>
          <Login />
        </ProvideAuth>
      </UserProvider>
    </MemoryRouter>
  );

  expect(container).toMatchSnapshot();
});
