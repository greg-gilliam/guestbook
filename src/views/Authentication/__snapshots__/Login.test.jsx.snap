// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the login page 1`] = `
<div>
  <fieldset>
    <legend>
      Sign in
    </legend>
    <form>
      <label
        for="username"
      >
        Username
      </label>
      <input
        id="username"
        name="username"
        type="text"
        value=""
      />
      <label
        for="password"
      >
        Password
      </label>
      <input
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="sign in"
        type="submit"
      >
        Sign in!
      </button>
    </form>
  </fieldset>
</div>
`;
